### ISSUE TITLE
Credo warns about missing spec when using @spec!

### ISSUE THREAD

**Issue by skwerlman** [Qqwy/elixir-type_check#102]
for this function head:
```elixir
  @spec! distance(String.t(), String.t()) :: integer()
  def distance(a, b)
```
credo gives this warning:
```elixir
  Code Readability
┃ 
┃ [R] → Functions should have a @spec type specification.
┃       lib/utils/damerau_levenshtein.ex:32:7 #(Utils.DamerauLevenshtein.distance)
```

**Comment by Qqwy** [comment]
Thank you very much for creating this issue!
This is definitely not intended behaviour. Which version of Credo are you using? And if you can share the rest of your module (not necessarily the code inside the `distance` function, but the signature and the surrounding code in the module including import statements) would really help to find out where this goes wrong.

**Comment by skwerlman** [comment]
Hi, here's the full code for the module: https://gitlab.tetrarch.co/commune_bot/utils/-/blob/scrybot/lib/utils/damerau_levenshtein.ex
Note that it currently has the specs check disabled so ci will pass

This is with credo 1.6.4

Lemme know if you have issues accessing it

**Comment by Qqwy** [comment]
Looking at [the source code of how Credo's check works](https://github.com/rrrene/credo/blob/master/lib/credo/check/readability/specs.ex) it seems like it looks at the source code as read by humans (before any compilation or macro expansion), and looking for `@spec` AST nodes in there.

This indeed won't work with TypeCheck, but it is not something that can be fixed as part of TypeCheck.

I guess we might need to write a Credo plugin which also recognizes `spec!`, possibly... :thinking:

**Comment by Qqwy** [comment]
Related: https://github.com/rrrene/credo/issues/883

In theory, a Credo plugin which uses `Code.Typespec.fetch_specs(ModuleUnderConsideration)` to receive a list of all functions for which specs exist in the given module (which can be cross-referenced with `ModuleUnderConsideration.__info__(:functions)`) should be able to do the trick.

**Comment by Qqwy** [comment]
@skwerlman Please try out the new check and see if it works well for you :blush:

**Comment by skwerlman** [comment]
Hi! Sorry for taking so long to get back, been busy moving.

Yes, the new check seems to work correctly! Thanks!
